### ISSUE TITLE
I am unable to detect torch.Tensor._version change during __torch_dispatch__ call

### ISSUE THREAD

**Issue by another-pjohnson** [pytorch/pytorch#64874]
## 🐛 Bug

I am unable to detect a version change using __torch_dispatch__ alone with a python subclassed torch.Tensor. Is there another hook/function after dispatch that would reflect the new updated version?

## To Reproduce

SCRIPT:

```
import torch
from torch.utils._pytree import tree_map

class DirtyTensorTracker(torch.Tensor):
    cpu_elem: torch.Tensor

    def __init__(self, cpu_elem: torch.Tensor):
        self.cpu_elem = cpu_elem
        self.is_dirty = False

    @staticmethod
    def __new__(cls, cpu_elem):
        r = torch.Tensor._make_subclass(cls, cpu_elem.to('cpu'), cpu_elem.requires_grad)
        return r

    #__torch_function__ = torch._C._disabled_torch_function_impl

    @classmethod
    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
        possibly_dirty = []
        def unwrap(e):
            if isinstance(e, DirtyTensorTracker):
                possibly_dirty.append((e, e.cpu_elem._version))
                return e.cpu_elem
            else:
                return e

        # need to test if tensor is modified

        r = func(*tree_map(unwrap, args), **tree_map(unwrap, kwargs))

        for e, saved_version in possibly_dirty:
            if saved_version != e.cpu_elem._version:
                print("detected dirty")
                e.is_dirty = True
        return r

x = DirtyTensorTracker(torch.randn((2, 2)))
x_old_version = x._version
x_cpu_elem_old_version = x.cpu_elem._version
# should be marked as dirty
x.add_(2)
x_new_version = x._version
x_cpu_elem_new_version = x.cpu_elem._version
print(f"x: old: {x_old_version} new: {x_new_version}")
print(f"x.cpu_elem: old: {x_cpu_elem_old_version} new: {x_cpu_elem_new_version}")
```

OUTPUT:
```
x: old: 0 new: 1
x.cpu_elem: old: 0 new: 1
is x dirty: False
```

## Expected behavior

OUTPUT:
```
x: old: 0 new: 1
x.cpu_elem: old: 0 new: 1
is x dirty: True
```

## Environment

 - PyTorch Version (e.g., 1.0): 1.10.0a0+git15eec8e
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): source into conda env
 - Build command you used (if compiling from source): 
 - Python version: 3.9.6
 - CUDA/cuDNN version: N/A
 - GPU models and configuration:N/A
 - Any other relevant information:N/A

## Additional context
This was discussed in a conversation with @ezyang about creating an OffloadTensor that could automatically move a tensor from disk to CPU for the execution of a function, then (if the tensor became dirty) write it back out to disk.

**Comment by ezyang** [comment]
The proximal cause for this problem is that `__torch_dispatch__` gets invoked after ADViewOrInplace which means that all version counter safeties are off at this point. This is actually a pretty serious problem because we're allowing generic user code. Maybe the correct thing to do is to move Python dispatch key between ADViewOrInplace and regular Autograd, not sure if this will work or not.

**Comment by albanD** [comment]
I think this is expected.
When working at the `__torch_dispatch__` level, you are "like a backend" meaning that you don't need to handle and don't have access to autograd-specific info.

In this case, you can just check from the function call what should happen.

**Comment by ezyang** [comment]
well except we don't have that info lol

**Comment by ezyang** [comment]
I tried YOLO'ing Python key before ADViewOrInplace and it mostly just broke stuff (no longer properly update VC when you do x.detach().add_(2))

**Comment by albanD** [comment]
These keys were not design to interpose anything in the middle for sure!
https://github.com/pytorch/pytorch/blob/85222c050f3fe2e2c7180f6f0aed5d0bbc7ab671/tools/autograd/gen_variable_type.py#L818

> well except we don't have that info lol

I think we should expose that for sure.
A but right now, `name[-1] == "_" or "out" in kwargs` should do the trick no?

**Comment by another-pjohnson** [comment]
@albanD I have a feeling that the condition should be a little more specific than that. For functions that end in _ (e.g. "add_") shouldn't we also check `e is args[0]`, for the "out" condition, shouldn't we do something similar? `out_is_this_tensor = False if "out" not in kwargs else e is kwargs["out"]`

**Comment by albanD** [comment]
Ho yes if you want to know which Tensor is the one modified.
cc @ezyang is the annotation plan going to tell you which arg is modified or just that one of them is?

**Comment by ezyang** [comment]
If I disambiguate overloads into torch dispatch (BC breaking!) I can give whole schema
